site/docs/skylark/windows_tips.md

```diff
@@ -0,0 +1,271 @@
+---
+layout: documentation
+title: Writing rules on Windows
+---
+
+# Writing rules on Windows
+
+Common problems of writing portable rules, and some solutions.
+
+This document focuses on writing Windows-compatible rules.
+
+## Paths
+
+Problems:
+
+- **Length limit**: maximum path length is 259 characters.
+
+  Though Windows also supports longer paths (up to 32767 characters), many programs are built with
+  the lower limit.
+
+  Be aware of this about programs you run in the actions.
+
+- **Working directory**: is also limited to 259 characters.
+
+  Processes cannot `cd` into a directory longer than 259 characters.
+
+- **Case-sensitivity**: Windows paths are case-insensitive, Unix paths are case-sensitive.
+
+  Be aware of this when creating command lines for actions.
+
+- **Path separators**: are backslash (`\`), not forward slash (`/`).
+
+  Bazel stores paths Unix-style, i.e. with `/` separators. Though some Windows programs support
+  Unix-style paths, others don't. Some built-in commands in cmd.exe support them, some don't.
+
+  It's best to always use `\` separators on Windows: replace `/` with `\` when you create command
+  lines and environment variables for actions.
+
+- **Absolute paths**: don't start with slash (`/`).
+
+  Absolute paths on Windows start with a drive letter, e.g. `C:\foo\bar.txt`. There's no single
+  filesystem root.
+
+  Be aware of this if your rule checks if a path is absolute. (Absolute paths should be avoided
+  though, they are often non-portable.)
+
+Solutions:
+
+- **Keep paths short.**
+
+  Avoid long directory names, deeply nested directory structures, long file names, long workspace
+  names, long target names.
+
+  All of these may become path components of actions' input files, and may exhaust the path length
+  limit.
+
+- **Use a short output root.**
+
+  Use the `--output_user_root=<path>` flag to specify a short path for Bazel outputs. A good idea
+  is to have a drive (or virtual drive) just for Bazel outputs (e.g. `D:\`), and adding this line
+  to your `.bazelrc` file:
+
+  ```
+  build --output_user_root=D:/
+  ```
+
+  or
+
+  ```
+  build --output_user_root=C:/_bzl
+  ```
+
+- **Use junctions.**
+
+  Junctions are, loosely speaking<sup>[1]</sup>, directory symlinks. Junctions are easy to create
+  and can point to directories (on the same computer) with long paths. If a build action creates a
+  junction whose path is short but whose target is long, then tools with short path limit can access
+  the files in the junction'ed directory.
+
+  In `.bat` files or in cmd.exe you can create junctions like so:
+
+  ```
+  mklink /J c:\path\to\junction c:\path\to\very\long\target\path
+  ```
+
+  <sup>[1]</sup>: Strictly speaking
+  [Junctions are not Symbolic Links](https://superuser.com/a/343079), but for sake of build actions
```

Comment:
Maybe we should have a individual section about "symlink", we can explain why runfiles tree doesn't exist by default on Windows, and what are the solutions (runfiles library / --enable_runfiles)

Reply:
Absolutely! Runfiles are a top TODO, see comments at the bottom of the file.

Reply:
Nice! I was reviewing the rendered markdown page, so I missed it.